File: frontend/src/lib/domain/encounterRulesets.ts
export const encounterRuleset2014 = "dnd-5e-2014-xp-v1";
export const encounterRuleset2024 = "dnd-5e-2024-xp-v1";

export type EncounterRuleset = typeof encounterRuleset2014 | typeof encounterRuleset2024;

export const encounterRulesetOptions = [
  { value: encounterRuleset2014, label: "2014 encounter rules" },
  { value: encounterRuleset2024, label: "2024 encounter rules" },
];

export function campaignEncounterRuleset(campaign: {
  allowedStandardSources: string[];
  encounterRuleset?: string;
}): EncounterRuleset {
  if (campaign.encounterRuleset === encounterRuleset2024) return encounterRuleset2024;
  if (campaign.encounterRuleset === encounterRuleset2014) return encounterRuleset2014;
  if (
    campaign.allowedStandardSources.includes("srd-5-2-1") &&
    !campaign.allowedStandardSources.includes("srd-2014")
  ) {
    return encounterRuleset2024;
  }
  return encounterRuleset2014;
}

export function encounterRulesetForSources(
  sources: string[],
  current: EncounterRuleset,
): EncounterRuleset {
  const has2014 = sources.includes("srd-2014");
  const has2024 = sources.includes("srd-5-2-1");
  if (has2024 && !has2014) return encounterRuleset2024;
  if (has2014 && !has2024) return encounterRuleset2014;
  return current;
}

export function difficultyOptions(ruleset: EncounterRuleset) {
  return ruleset === encounterRuleset2024
    ? [
        { label: "Low", value: "low" },
        { label: "Moderate", value: "moderate" },
        { label: "High", value: "high" },
      ]
    : [
        { label: "Easy", value: "easy" },
        { label: "Medium", value: "medium" },
        { label: "Hard", value: "hard" },
        { label: "Deadly", value: "deadly" },
      ];
}

export function defaultDifficulty(ruleset: EncounterRuleset) {
  return ruleset === encounterRuleset2024 ? "moderate" : "medium";
}

File: backend/internal/app/encounter_rulesets.go
package app

import (
	"strings"

	dbmodels "bludm/backend/internal/db"
	"bludm/backend/internal/generation"
	"bludm/backend/internal/models"
	"bludm/backend/internal/rulesets"
)

func campaignEncounterRuleset(campaign models.Campaign) (string, error) {
	resolved, err := rulesets.ResolveEncounterRuleset(
		campaign.AllowedStandardSources, campaign.EncounterRuleset,
	)
	if err != nil {
		return "", ValidationError(
			"encounter_ruleset_required",
			err.Error(),
			map[string]any{"allowedStandardSources": campaign.AllowedStandardSources},
		)
	}
	return resolved, nil
}

func persistedEncounterRuleset(
	encounter dbmodels.EncounterEntity,
	campaign models.Campaign,
) (string, error) {
	if rulesets.IsEncounterRuleset(encounter.DifficultyRuleset) {
		return encounter.DifficultyRuleset, nil
	}
	if value, ok := encounter.Metadata["difficultyRuleset"].(string); ok &&
		rulesets.IsEncounterRuleset(strings.TrimSpace(value)) {
		return strings.TrimSpace(value), nil
	}
	if evidence, ok := encounter.Metadata["difficultyEvidence"].(map[string]any); ok {
		if value, ok := evidence["ruleset"].(string); ok &&
			rulesets.IsEncounterRuleset(strings.TrimSpace(value)) {
			return strings.TrimSpace(value), nil
		}
	}
	return campaignEncounterRuleset(campaign)
}

func normalizeDifficultyForRuleset(ruleset, value string) (string, error) {
	if normalized, ok := generation.NormalizeDifficulty(ruleset, value); ok {
		return normalized, nil
	}
	message := "difficulty must be easy, medium, hard, or deadly for 2014 rules"
	if ruleset == rulesets.Encounter2024 {
		message = "difficulty must be low, moderate (or medium), or high for 2024 rules"
	}
	return "", ValidationError("invalid_difficulty", message, map[string]any{"ruleset": ruleset})
}
